src/components/Sketches/components/EditSketchModal.js

```diff
@@ -0,0 +1,323 @@
+      if (
+        this.state.newLanguage.display === LanguageDropdownValues[i].display &&
+        this.state.newLanguage.value === LanguageDropdownValues[i].value
+      ) {
+        notFound = false;
+        break;
+      }
+    }
+    if (notFound) {
+      this.setState({ error: "Invalid language selected" });
+      return true;
+    }
+
+    return false;
+  };
+
+  handleSubmitEdit = async e => {
+    e.preventDefault();
+
+    if (this.badNameInput() || this.badLanguageInput()) {
+      // note it's impossible to have a bad thumbnail input
+      return;
+    }
+
+    let data = {};
+    let doUpdate = false;
+
+    if (this.state.newLanguage !== -1) {
+      data["language"] = this.state.newLanguage.value;
+      doUpdate = true;
+    }
+    if (this.state.newName !== -1) {
+      data["name"] = this.state.newName;
+      doUpdate = true;
+    }
+    if (this.state.newThumbnail !== -1) {
+      data["thumbnail"] = this.state.newThumbnail;
+      doUpdate = true;
+    }
+    if (doUpdate) {
+      let updateData = {};
+      updateData[this.props.sketchKey] = data;
+      try {
+        fetch
+          .updatePrograms(this.props.uid, updateData)
+          .then(res => {
+            return res.json();
+          })
+          .then(json => {
+            if (!json.ok) {
+              this.setState({
+                spinner: false,
+                error: json.error || "Failed to create sketch, please try again later",
+              });
+              return;
+            }
+            if (this.state.newLanguage !== -1) {
+              this.props.setProgramLanguage(this.props.sketchKey, this.state.newLanguage.value);
+            }
+            if (this.state.newName !== -1) {
+              this.props.setProgramName(this.props.sketchKey, this.state.newName);
+            }
+            if (this.state.newThumbnail !== -1) {
+              this.props.setProgramThumbnail(this.props.sketchKey, this.state.newThumbnail);
+            }
+            this.closeModal();
+          })
+          .catch(err => {
+            this.setState({
+              spinner: false,
+              error: "Failed to create sketch, please try again later",
+            });
+            console.log(err);
+          });
+      } catch (err) {
+        console.log(err);
+      }
+      this.setState({ spinner: true, error: "" });
+    }
+  };
+
+  renderMainModal() {
+    let thumbnailPreview =
+      this.state.newThumbnail !== -1 ? (
+        <img
+          src={`${process.env.PUBLIC_URL}/img/sketch-thumbnails/${
+            SketchThumbnailArray[this.state.newThumbnail]
+          }.svg`}
+          className={"sketches-modal-header-thumbnail"}
+          alt="icon"
+        />
+      ) : (
+        <img
+          src={`${process.env.PUBLIC_URL}/img/sketch-thumbnails/${this.props.sketchImg}.svg`}
+          className={"sketches-modal-header-thumbnail"}
+          alt="icodn"
+        />
+      );
+    return (
+      <ReactModal
```

Comment:
I know you're not a fan of the renderBlah functions, but imo, they improve comprehension of these big jsx's.

Reply:
Reasonable, what should I break up? Probably each of the inputs, right?

Reply:
We can have a big ol readability change where we make everything consistent. We can talk bout it next week. For now, it's fine